File: Sources/VIPSImage.m
//
//  VIPSImage.m
//  VIPSKit
//
//  Core implementation: initialization, properties, memory management
//

#import "VIPSImage+Private.h"
#import <objc/runtime.h>
#import <objc/message.h>

NSString *const VIPSErrorDomain = @"org.libvips.VIPSKit";

#pragma mark - VIPSImageStatistics

@interface VIPSImageStatistics ()
@property (nonatomic, readwrite) double min;
@property (nonatomic, readwrite) double max;
@property (nonatomic, readwrite) double mean;
@property (nonatomic, readwrite) double standardDeviation;
@end

@implementation VIPSImageStatistics
@end

#pragma mark - VIPSImage

@implementation VIPSImage

#pragma mark - Error Handling

+ (NSError *)errorFromVips {
    const char *msg = vips_error_buffer();
    NSString *message = msg ? [NSString stringWithUTF8String:msg] : @"Unknown VIPS error";
    vips_error_clear();
    return [NSError errorWithDomain:VIPSErrorDomain
                               code:-1
                           userInfo:@{NSLocalizedDescriptionKey: message}];
}

#pragma mark - Lifecycle

+ (BOOL)initializeWithError:(NSError **)error {
    if (VIPS_INIT("VIPSKit") != 0) {
        if (error) {
            *error = [self errorFromVips];
        }
        return NO;
    }

    // Configure cache for typical usage.
    vips_cache_set_max(100);
    vips_cache_set_max_mem(50 * 1024 * 1024);  // 50MB
    vips_cache_set_max_files(10);

    // Default to single-threaded for batch processing
    // (parallelize across images, not within single image)
    vips_concurrency_set(1);

    return YES;
}

+ (void)shutdown {
    vips_shutdown();
}

- (instancetype)initWithVipsImage:(VipsImage *)image {
    self = [super init];
    if (self) {
        _image = image;
        g_object_ref(image);
    }
    return self;
}

- (void)dealloc {
    if (_image) {
        g_object_unref(_image);
    }
}

#pragma mark - Properties

- (NSInteger)width {
    return vips_image_get_width(self.image);
}

- (NSInteger)height {
    return vips_image_get_height(self.image);
}

- (NSInteger)bands {
    return vips_image_get_bands(self.image);
}

- (BOOL)hasAlpha {
    return vips_image_hasalpha(self.image);
}

- (NSString *)loaderName {
    const char *loader = NULL;
    if (vips_image_get_string(self.image, VIPS_META_LOADER, &loader) == 0 && loader) {
        return [NSString stringWithUTF8String:loader];
    }
    return nil;
}

- (VIPSImageFormat)sourceFormat {
    NSString *loader = self.loaderName;
    if (!loader) {
        return VIPSImageFormatUnknown;
    }

    if ([loader hasPrefix:@"jpeg"] || [loader hasPrefix:@"jpg"]) {
        return VIPSImageFormatJPEG;
    } else if ([loader hasPrefix:@"png"]) {
        return VIPSImageFormatPNG;
    } else if ([loader hasPrefix:@"webp"]) {
        return VIPSImageFormatWebP;
    } else if ([loader hasPrefix:@"heif"]) {
        // Check for AVIF specifically - heifload handles both
        const char *compression = NULL;
        if (vips_image_get_string(self.image, "heif-compression", &compression) == 0 && compression) {
            if (strcmp(compression, "av1") == 0) {
                return VIPSImageFormatAVIF;
            }
        }
        return VIPSImageFormatHEIF;
    } else if ([loader hasPrefix:@"jxl"]) {
        return VIPSImageFormatJXL;
    } else if ([loader hasPrefix:@"gif"]) {
        return VIPSImageFormatGIF;
    }

    return VIPSImageFormatUnknown;
}

#pragma mark - Class Memory Management

+ (void)setCacheMaxOperations:(NSInteger)max {
    vips_cache_set_max((int)max);
}

+ (void)setCacheMaxMemory:(NSInteger)bytes {
    vips_cache_set_max_mem((size_t)bytes);
}

+ (void)setCacheMaxFiles:(NSInteger)max {
    vips_cache_set_max_files((int)max);
}

+ (void)clearCache {
    // Don't use vips_cache_drop_all() - it destroys the hash table entirely
    // and sets vips_cache_table to NULL, causing crashes on subsequent operations.
    //
    // Instead, use the safe approach: temporarily set max to 0 to evict all
    // cached operations via the normal LRU mechanism, then restore the limit.
    int originalMax = vips_cache_get_max();
    vips_cache_set_max(0);  // Triggers vips_cache_trim() which evicts everything
    vips_cache_set_max(originalMax);  // Restore
}

+ (NSInteger)memoryUsage {
    return (NSInteger)vips_tracked_get_mem();
}

+ (NSInteger)memoryHighWater {
    return (NSInteger)vips_tracked_get_mem_highwater();
}

+ (void)resetMemoryHighWater {
    vips_tracked_get_mem_highwater();
}

+ (void)setConcurrency:(NSInteger)threads {
    vips_concurrency_set((int)threads);
}

+ (NSInteger)concurrency {
    return (NSInteger)vips_concurrency_get();
}

#pragma mark - Instance Memory Management

- (VIPSImage *)copyToMemoryWithError:(NSError **)error {
    VipsImage *out = vips_image_copy_memory(self.image);

    if (!out) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    VIPSImage *result = [[VIPSImage alloc] init];
    result.image = out;
    return result;
}

#pragma mark - Pixel Access

- (BOOL)withPixelData:(void (NS_NOESCAPE ^)(const uint8_t *data,
                                            NSInteger width,
                                            NSInteger height,
                                            NSInteger bytesPerRow,
                                            NSInteger bands))block
                error:(NSError **)error {
    if (!block) {
        return YES;
    }

    VipsImage *prepared = NULL;

    // Convert to sRGB if needed
    VipsInterpretation interpretation = vips_image_get_interpretation(self.image);
    if (interpretation != VIPS_INTERPRETATION_sRGB &&
        interpretation != VIPS_INTERPRETATION_RGB &&
        interpretation != VIPS_INTERPRETATION_B_W) {
        if (vips_colourspace(self.image, &prepared, VIPS_INTERPRETATION_sRGB, NULL) != 0) {
            if (error) {
                *error = [self.class errorFromVips];
            }
            return NO;
        }
    } else {
        prepared = self.image;
        g_object_ref(prepared);
    }

    // Ensure 8-bit format
    if (vips_image_get_format(prepared) != VIPS_FORMAT_UCHAR) {
        VipsImage *cast = NULL;
        if (vips_cast_uchar(prepared, &cast, NULL) != 0) {
            g_object_unref(prepared);
            if (error) {
                *error = [self.class errorFromVips];
            }
            return NO;
        }
        g_object_unref(prepared);
        prepared = cast;
    }

    // Get dimensions
    int width = vips_image_get_width(prepared);
    int height = vips_image_get_height(prepared);
    int bands = vips_image_get_bands(prepared);
    size_t bytesPerRow = (size_t)width * (size_t)bands;

    // Write to contiguous memory buffer (single copy from vips pipeline)
    size_t dataSize = 0;
    void *data = vips_image_write_to_memory(prepared, &dataSize);
    g_object_unref(prepared);

    if (!data) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return NO;
    }

    // Call block with pixel data
    block((const uint8_t *)data, width, height, (NSInteger)bytesPerRow, bands);

    // Free the pixel data
    g_free(data);

    return YES;
}

#pragma mark - Analysis

- (CGRect)findTrimWithError:(NSError **)error {
    return [self findTrimWithThreshold:10.0 background:nil error:error];
}

- (CGRect)findTrimWithThreshold:(double)threshold error:(NSError **)error {
    return [self findTrimWithThreshold:threshold background:nil error:error];
}

- (CGRect)findTrimWithThreshold:(double)threshold
                     background:(NSArray<NSNumber *> *)background
                          error:(NSError **)error {
    int left = 0, top = 0, width = 0, height = 0;
    int result;

    if (background && background.count > 0) {
        // Build VipsArrayDouble from the background array
        double *bgValues = g_malloc(sizeof(double) * background.count);
        for (NSUInteger i = 0; i < background.count; i++) {
            bgValues[i] = background[i].doubleValue;
        }
        VipsArrayDouble *bgArray = vips_array_double_new(bgValues, (int)background.count);
        g_free(bgValues);

        result = vips_find_trim(self.image, &left, &top, &width, &height,
                                "threshold", threshold,
                                "background", bgArray,
                                NULL);
        vips_area_unref(VIPS_AREA(bgArray));
    } else {
        // Auto-detect background
        result = vips_find_trim(self.image, &left, &top, &width, &height,
                                "threshold", threshold,
                                NULL);
    }

    if (result != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return CGRectNull;
    }

    return CGRectMake(left, top, width, height);
}

- (VIPSImageStatistics *)statisticsWithError:(NSError **)error {
    double min = 0, max = 0, mean = 0, stddev = 0;

    if (vips_min(self.image, &min, NULL) != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    if (vips_max(self.image, &max, NULL) != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    if (vips_avg(self.image, &mean, NULL) != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    if (vips_deviate(self.image, &stddev, NULL) != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    VIPSImageStatistics *stats = [[VIPSImageStatistics alloc] init];
    stats.min = min;
    stats.max = max;
    stats.mean = mean;
    stats.standardDeviation = stddev;
    return stats;
}

- (NSArray<NSNumber *> *)averageColorWithError:(NSError **)error {
    VipsImage *stats = NULL;

    // vips_stats returns a (bands+1) x 10 image:
    // - Columns: 0=all bands combined, 1-n=individual bands
    // - Rows: 0=min, 1=max, 2=sum, 3=sum², 4=mean, 5=stddev, 6-9=position info
    if (vips_stats(self.image, &stats, NULL) != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    int bands = vips_image_get_bands(self.image);
    NSMutableArray<NSNumber *> *result = [NSMutableArray arrayWithCapacity:bands];

    // Read mean values (row 4) for each band (columns 1 through bands)
    for (int band = 1; band <= bands; band++) {
        double *pixel = (double *)vips_image_get_data(stats);
        if (!pixel) {
            g_object_unref(stats);
            if (error) {
                *error = [NSError errorWithDomain:VIPSErrorDomain
                                             code:-1
                                         userInfo:@{NSLocalizedDescriptionKey: @"Failed to read stats data"}];
            }
            return nil;
        }
        // Stats image is (bands+1) columns wide, row 4 is the mean
        int statsWidth = bands + 1;
        double mean = pixel[4 * statsWidth + band];
        [result addObject:@(mean)];
    }

    g_object_unref(stats);
    return [result copy];
}

- (NSArray<NSNumber *> *)detectBackgroundColorWithError:(NSError **)error {
    return [self detectBackgroundColorWithStripWidth:10 error:error];
}

- (NSArray<NSNumber *> *)detectBackgroundColorWithStripWidth:(NSInteger)stripWidth error:(NSError **)error {
    if (stripWidth <= 0) {
        stripWidth = 10;
    }

    NSInteger w = self.width;
    NSInteger h = self.height;

    // Handle very small images
    if (w <= stripWidth * 2 || h <= stripWidth * 2) {
        // Image is too small, just return average of whole image
        return [self averageColorWithError:error];
    }

    // Extract edge strips and combine them
    // We'll create a composite of all edge pixels
    VipsImage *top = NULL, *bottom = NULL, *left = NULL, *right = NULL;
    VipsImage *horizontal = NULL, *vertical = NULL, *combined = NULL;

    // Top strip (full width)
    if (vips_crop(self.image, &top, 0, 0, (int)w, (int)stripWidth, NULL) != 0) {
        if (error) *error = [self.class errorFromVips];
        return nil;
    }

    // Bottom strip (full width)
    if (vips_crop(self.image, &bottom, 0, (int)(h - stripWidth), (int)w, (int)stripWidth, NULL) != 0) {
        g_object_unref(top);
        if (error) *error = [self.class errorFromVips];
        return nil;
    }

    // Left strip (excluding corners already covered by top/bottom)
    if (vips_crop(self.image, &left, 0, (int)stripWidth, (int)stripWidth, (int)(h - 2 * stripWidth), NULL) != 0) {
        g_object_unref(top);
        g_object_unref(bottom);
        if (error) *error = [self.class errorFromVips];
        return nil;
    }

    // Right strip (excluding corners)
    if (vips_crop(self.image, &right, (int)(w - stripWidth), (int)stripWidth, (int)stripWidth, (int)(h - 2 * stripWidth), NULL) != 0) {
        g_object_unref(top);
        g_object_unref(bottom);
        g_object_unref(left);
        if (error) *error = [self.class errorFromVips];
        return nil;
    }

    // Join top and bottom vertically
    if (vips_join(top, bottom, &horizontal, VIPS_DIRECTION_VERTICAL, NULL) != 0) {
        g_object_unref(top);
        g_object_unref(bottom);
        g_object_unref(left);
        g_object_unref(right);
        if (error) *error = [self.class errorFromVips];
        return nil;
    }
    g_object_unref(top);
    g_object_unref(bottom);

    // Join left and right vertically
    if (vips_join(left, right, &vertical, VIPS_DIRECTION_VERTICAL, NULL) != 0) {
        g_object_unref(horizontal);
        g_object_unref(left);
        g_object_unref(right);
        if (error) *error = [self.class errorFromVips];
        return nil;
    }
    g_object_unref(left);
    g_object_unref(right);

    // Join horizontal and vertical strips
    if (vips_join(horizontal, vertical, &combined, VIPS_DIRECTION_VERTICAL, NULL) != 0) {
        g_object_unref(horizontal);
        g_object_unref(vertical);
        if (error) *error = [self.class errorFromVips];
        return nil;
    }
    g_object_unref(horizontal);
    g_object_unref(vertical);

    // Create a temporary VIPSImage wrapper and get its average color
    VIPSImage *combinedWrapper = [[VIPSImage alloc] init];
    combinedWrapper.image = combined;  // Takes ownership

    return [combinedWrapper averageColorWithError:error];
}

#pragma mark - Arithmetic

- (VIPSImage *)subtract:(VIPSImage *)image error:(NSError **)error {
    VipsImage *out = NULL;

    if (vips_subtract(self.image, image.image, &out, NULL) != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    VIPSImage *result = [[VIPSImage alloc] init];
    result.image = out;
    return result;
}

- (VIPSImage *)absoluteWithError:(NSError **)error {
    VipsImage *out = NULL;

    if (vips_abs(self.image, &out, NULL) != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    VIPSImage *result = [[VIPSImage alloc] init];
    result.image = out;
    return result;
}

#pragma mark - Debug Support

- (id)debugQuickLookObject {
    // Xcode calls this method to show image previews in the debugger
    CGImageRef cgImage = [self createCGImageWithError:nil];
    if (!cgImage) {
        return [NSString stringWithFormat:@"VIPSImage %ldx%ld (%ld bands)",
                (long)self.width, (long)self.height, (long)self.bands];
    }

    // Use runtime lookup for UIImage to avoid compile-time UIKit dependency
    // UIKit is linked at runtime but headers may not be available during build
    Class UIImageClass = NSClassFromString(@"UIImage");
    if (UIImageClass) {
        // Use objc_msgSend to call +[UIImage imageWithCGImage:]
        SEL selector = NSSelectorFromString(@"imageWithCGImage:");
        if ([UIImageClass respondsToSelector:selector]) {
            // Cast objc_msgSend to the correct function signature
            id (*msgSend)(Class, SEL, CGImageRef) = (id (*)(Class, SEL, CGImageRef))objc_msgSend;
            id image = msgSend(UIImageClass, selector, cgImage);
            CGImageRelease(cgImage);
            return image;
        }
    }

    CGImageRelease(cgImage);
    return [NSString stringWithFormat:@"VIPSImage %ldx%ld (%ld bands)",
            (long)self.width, (long)self.height, (long)self.bands];
}

@end

File: Sources/VIPSImage.swift
import Foundation
import CoreGraphics
internal import vips
internal import CVIPS

#if canImport(UIKit)
import UIKit
#elseif canImport(AppKit)
import AppKit
#endif

/// A high-performance image processing class powered by libvips.
///
/// `VIPSImage` wraps the libvips C library to provide fast, memory-efficient image
/// operations suitable for batch processing and thumbnail generation. Each instance
/// represents an immutable image — all operations return new `VIPSImage` instances.
///
/// You must call ``initialize()`` once before using any other VIPSKit functionality.
///
/// ```swift
/// try VIPSImage.initialize()
/// let image = try VIPSImage(contentsOfFile: "/path/to/photo.jpg")
/// let thumbnail = try image.resizeToFit(width: 200, height: 200)
/// let cgImage = try thumbnail.cgImage
/// ```
public final class VIPSImage: @unchecked Sendable {

    /// The underlying libvips image pointer.
    internal let pointer: UnsafeMutablePointer<VipsImage>

    // MARK: - Lifecycle

    /// Wrap an existing VipsImage pointer, taking ownership of the reference.
    /// The caller must not release the pointer after passing it to this initializer.
    /// - Parameter pointer: A VipsImage pointer to take ownership of
    internal init(pointer: UnsafeMutablePointer<VipsImage>) {
        self.pointer = pointer
    }

    /// Wrap an existing VipsImage pointer by adding a new reference.
    /// The caller retains their own reference and must release it independently.
    /// - Parameter pointer: A VipsImage pointer to borrow
    internal init(borrowing pointer: UnsafeMutablePointer<VipsImage>) {
        g_object_ref(gpointer(pointer))
        self.pointer = pointer
    }

    deinit {
        g_object_unref(gpointer(pointer))
    }

    // MARK: - Initialization

    /// Initialize the libvips library. This must be called once before
    /// performing any image operations.
    ///
    /// Configures default cache and concurrency settings optimized for batch processing:
    /// - Operation cache: 100 operations, 50MB, 10 open files
    /// - Concurrency: 1 thread (parallelize at the application layer instead)
    public static func initialize() throws {
        if vips_init("VIPSKit") != 0 {
            throw VIPSError.fromVips()
        }
        vips_cache_set_max(100)
        vips_cache_set_max_mem(50 * 1024 * 1024) // 50MB
        vips_cache_set_max_files(10)
        vips_concurrency_set(1)
    }

    /// Shut down the libvips library and release all associated resources.
    /// Call this at application termination if desired. Optional.
    public static func shutdown() {
        vips_shutdown()
    }

    // MARK: - Properties

    /// The width of the image in pixels.
    public var width: Int { Int(vips_image_get_width(pointer)) }

    /// The height of the image in pixels.
    public var height: Int { Int(vips_image_get_height(pointer)) }

    /// The number of bands (channels) in the image. For example, 3 for RGB and 4 for RGBA.
    public var bands: Int { Int(vips_image_get_bands(pointer)) }

    /// Whether the image contains an alpha transparency channel.
    public var hasAlpha: Bool { vips_image_hasalpha(pointer) != 0 }

    /// The internal loader name used by libvips when this image was loaded
    /// (e.g., `"jpegload"`, `"pngload"`), or `nil` if the image was not loaded from a file or buffer.
    public var loaderName: String? {
        var loader: UnsafePointer<CChar>?
        guard vips_image_get_string(pointer, VIPS_META_LOADER, &loader) == 0,
              let loader else { return nil }
        return String(cString: loader)
    }

    /// The detected source format of the image based on the loader used.
    /// Returns ``VIPSImageFormat/unknown`` if the format could not be determined.
    public var sourceFormat: VIPSImageFormat {
        guard let loader = loaderName else { return .unknown }
        if loader.hasPrefix("jpeg") || loader.hasPrefix("jpg") {
            return .jpeg
        } else if loader.hasPrefix("png") {
            return .png
        } else if loader.hasPrefix("webp") {
            return .webP
        } else if loader.hasPrefix("heif") {
            var compression: UnsafePointer<CChar>?
            if vips_image_get_string(pointer, "heif-compression", &compression) == 0,
               let compression, strcmp(compression, "av1") == 0 {
                return .avif
            }
            return .heif
        } else if loader.hasPrefix("jxl") {
            return .jxl
        } else if loader.hasPrefix("gif") {
            return .gif
        } else if loader.hasPrefix("tiff") {
            return .tiff
        }
        return .unknown
    }

    // MARK: - Cache

    /// Controls for the libvips operation cache.
    public struct Cache {
        /// The maximum number of operations to keep in the cache.
        /// Set to 0 to disable caching entirely.
        public static var maxOperations: Int {
            get { Int(vips_cache_get_max()) }
            set { vips_cache_set_max(Int32(newValue)) }
        }

        /// The maximum amount of memory used by the operation cache, in bytes.
        public static var maxMemory: Int {
            get { vips_cache_get_max_mem() }
            set { vips_cache_set_max_mem(newValue) }
        }

        /// The maximum number of open files held by the cache.
        public static var maxFiles: Int {
            get { Int(vips_cache_get_max_files()) }
            set { vips_cache_set_max_files(Int32(newValue)) }
        }

        /// Clear all cached operations and free associated memory.
        /// Uses a safe workaround that avoids the `vips_cache_drop_all()` crash.
        public static func clear() {
            let originalMax = vips_cache_get_max()
            vips_cache_set_max(0)
            vips_cache_set_max(originalMax)
        }
    }

    /// The current memory usage tracked by libvips in bytes.
    public static var memoryUsage: Int { Int(vips_tracked_get_mem()) }

    /// The peak memory usage tracked by libvips since the last reset, in bytes.
    public static var memoryHighWater: Int { Int(vips_tracked_get_mem_highwater()) }

    /// Reset the peak memory tracking counter back to the current usage level.
    public static func resetMemoryHighWater() {
        _ = vips_tracked_get_mem_highwater()
    }

    /// The number of worker threads used by libvips for internal parallelism.
    /// Set to 0 to auto-detect based on available CPU cores.
    public static var concurrency: Int {
        get { Int(vips_concurrency_get()) }
        set { vips_concurrency_set(Int32(newValue)) }
    }

    // MARK: - Instance Memory Management

    /// Copy the image pixels into a new contiguous memory block, breaking
    /// any lazy evaluation chain. This allows source images to be freed
    /// even if downstream images still exist.
    /// - Returns: A new image with all pixels evaluated and stored in memory
    public func copiedToMemory() throws -> VIPSImage {
        guard let out = vips_image_copy_memory(pointer) else {
            throw VIPSError.fromVips()
        }
        return VIPSImage(pointer: out)
    }

    // MARK: - Pixel Access

    /// A snapshot of raw pixel data, valid only within the `withPixelData` closure.
    public struct PixelBuffer {
        /// A pointer to the raw pixel bytes.
        public let data: UnsafePointer<UInt8>
        /// The image width in pixels.
        public let width: Int
        /// The image height in pixels.
        public let height: Int
        /// The number of bytes per row (stride).
        public let bytesPerRow: Int
        /// The number of bands (3 for RGB, 4 for RGBA).
        public let bands: Int
    }

    /// Provides zero-copy access to the image's raw pixel data within a closure.
    ///
    /// The pixel data is converted to 8-bit unsigned sRGB format before being passed
    /// to the closure. The ``PixelBuffer`` is only valid within the closure's scope.
    ///
    /// - Parameter body: A closure that receives a ``PixelBuffer`` with the pixel data
    /// - Returns: The value returned by the closure
    public func withPixelData<T>(_ body: (PixelBuffer) throws -> T) throws -> T {
        var prepared: UnsafeMutablePointer<VipsImage>

        let interpretation = vips_image_get_interpretation(pointer)
        if interpretation != VIPS_INTERPRETATION_sRGB &&
           interpretation != VIPS_INTERPRETATION_RGB &&
           interpretation != VIPS_INTERPRETATION_B_W {
            var srgb: UnsafeMutablePointer<VipsImage>?
            guard cvips_colourspace(pointer, &srgb, VIPS_INTERPRETATION_sRGB) == 0,
                  let srgb else { throw VIPSError.fromVips() }
            prepared = srgb
        } else {
            g_object_ref(gpointer(pointer))
            prepared = pointer
        }

        if vips_image_get_format(prepared) != VIPS_FORMAT_UCHAR {
            var cast: UnsafeMutablePointer<VipsImage>?
            guard cvips_cast_uchar(prepared, &cast) == 0, let cast else {
                g_object_unref(gpointer(prepared))
                throw VIPSError.fromVips()
            }
            g_object_unref(gpointer(prepared))
            prepared = cast
        }

        let w = Int(vips_image_get_width(prepared))
        let h = Int(vips_image_get_height(prepared))
        let b = Int(vips_image_get_bands(prepared))
        let bytesPerRow = w * b

        var dataSize: Int = 0
        guard let data = vips_image_write_to_memory(prepared, &dataSize) else {
            g_object_unref(gpointer(prepared))
            throw VIPSError.fromVips()
        }
        g_object_unref(gpointer(prepared))

        defer { g_free(data) }
        let buffer = PixelBuffer(data: data.assumingMemoryBound(to: UInt8.self),
                                 width: w, height: h, bytesPerRow: bytesPerRow, bands: b)
        return try body(buffer)
    }

}

// MARK: - Debug Description

extension VIPSImage: CustomDebugStringConvertible {

    public var debugDescription: String {
        var lines: [String] = []
        lines.append("VIPSImage: \(width)x\(height) pixels, \(bands) band\(bands == 1 ? "" : "s")\(hasAlpha ? " (with alpha)" : "")")

        // Source format and loader
        let format = sourceFormat
        if format != .unknown {
            lines.append("  Format: \(format.debugLabel)\(loaderName.map { " (loader: \($0))" } ?? "")")
        } else if let loader = loaderName {
            lines.append("  Loader: \(loader)")
        }

        // Pixel format and interpretation
        let interp = vips_image_get_interpretation(pointer)
        let bandFmt = vips_image_get_format(pointer)
        lines.append("  Interpretation: \(Self.interpretationLabel(interp))")
        lines.append("  Band format: \(Self.bandFormatLabel(bandFmt))")

        // Resolution (only if non-default)
        let xres = xResolution
        let yres = yResolution
        if xres > 0 || yres > 0 {
            let dpiX = xres * 25.4
            let dpiY = yres * 25.4
            if abs(dpiX - dpiY) < 0.01 {
                lines.append("  Resolution: \(String(format: "%.1f", dpiX)) DPI")
            } else {
                lines.append("  Resolution: \(String(format: "%.1f", dpiX)) x \(String(format: "%.1f", dpiY)) DPI")
            }
        }

        // Orientation
        if let orient = orientation, orient != 1 {
            lines.append("  Orientation: \(orient) (\(Self.orientationLabel(orient)))")
        }

        // Multi-page
        let pages = pageCount
        if pages > 1 {
            lines.append("  Pages: \(pages)\(pageHeight.map { ", page height: \($0)px" } ?? "")")
        }

        // Embedded data
        var embeds: [String] = []
        if exifData != nil { embeds.append("EXIF") }
        if xmpData != nil { embeds.append("XMP") }
        if iccProfile != nil { embeds.append("ICC") }
        if !embeds.isEmpty {
            lines.append("  Embedded data: \(embeds.joined(separator: ", "))")
        }

        // Memory
        let estimatedBytes = width * height * bands * Self.bytesPerBand(bandFmt)
        lines.append("  Estimated size: \(Self.formatBytes(estimatedBytes))")

        return lines.joined(separator: "\n")
    }

    // MARK: - Debug Label Helpers

    private static func interpretationLabel(_ interp: VipsInterpretation) -> String {
        switch interp {
        case VIPS_INTERPRETATION_sRGB:       return "sRGB"
        case VIPS_INTERPRETATION_RGB:        return "RGB (linear)"
        case VIPS_INTERPRETATION_RGB16:      return "RGB16"
        case VIPS_INTERPRETATION_B_W:        return "Grayscale"
        case VIPS_INTERPRETATION_GREY16:     return "Grayscale 16-bit"
        case VIPS_INTERPRETATION_CMYK:       return "CMYK"
        case VIPS_INTERPRETATION_LAB:        return "CIE Lab"
        case VIPS_INTERPRETATION_LABS:       return "CIE LabS"
        case VIPS_INTERPRETATION_LCH:        return "CIE LCh"
        case VIPS_INTERPRETATION_XYZ:        return "CIE XYZ"
        case VIPS_INTERPRETATION_scRGB:      return "scRGB (linear)"
        case VIPS_INTERPRETATION_HSV:        return "HSV"
        case VIPS_INTERPRETATION_MULTIBAND:  return "Multiband"
        case VIPS_INTERPRETATION_FOURIER:    return "Fourier"
        case VIPS_INTERPRETATION_MATRIX:     return "Matrix"
        default:                             return "Unknown (\(interp.rawValue))"
        }
    }

    private static func bandFormatLabel(_ fmt: VipsBandFormat) -> String {
        switch fmt {
        case VIPS_FORMAT_UCHAR:    return "8-bit unsigned"
        case VIPS_FORMAT_CHAR:     return "8-bit signed"
        case VIPS_FORMAT_USHORT:   return "16-bit unsigned"
        case VIPS_FORMAT_SHORT:    return "16-bit signed"
        case VIPS_FORMAT_UINT:     return "32-bit unsigned"
        case VIPS_FORMAT_INT:      return "32-bit signed"
        case VIPS_FORMAT_FLOAT:    return "32-bit float"
        case VIPS_FORMAT_DOUBLE:   return "64-bit float"
        case VIPS_FORMAT_COMPLEX:  return "64-bit complex"
        case VIPS_FORMAT_DPCOMPLEX: return "128-bit complex"
        default:                   return "Unknown (\(fmt.rawValue))"
        }
    }

    private static func orientationLabel(_ value: Int) -> String {
        switch value {
        case 1: return "normal"
        case 2: return "flipped horizontal"
        case 3: return "rotated 180°"
        case 4: return "flipped vertical"
        case 5: return "transposed"
        case 6: return "rotated 90° CW"
        case 7: return "transverse"
        case 8: return "rotated 270° CW"
        default: return "unknown"
        }
    }

    private static func bytesPerBand(_ fmt: VipsBandFormat) -> Int {
        switch fmt {
        case VIPS_FORMAT_UCHAR, VIPS_FORMAT_CHAR: return 1
        case VIPS_FORMAT_USHORT, VIPS_FORMAT_SHORT: return 2
        case VIPS_FORMAT_UINT, VIPS_FORMAT_INT, VIPS_FORMAT_FLOAT: return 4
        case VIPS_FORMAT_DOUBLE, VIPS_FORMAT_COMPLEX: return 8
        case VIPS_FORMAT_DPCOMPLEX: return 16
        default: return 1
        }
    }

    private static func formatBytes(_ bytes: Int) -> String {
        if bytes < 1024 { return "\(bytes) B" }
        let kb = Double(bytes) / 1024
        if kb < 1024 { return String(format: "%.1f KB", kb) }
        let mb = kb / 1024
        if mb < 1024 { return String(format: "%.1f MB", mb) }
        let gb = mb / 1024
        return String(format: "%.2f GB", gb)
    }
}

// MARK: - Debugger Quick Look

extension VIPSImage {
    @objc func debugQuickLookObject() -> Any? {
        guard let cgImage = try? self.cgImage else {
            return "VIPSImage(\(width)x\(height), \(bands) bands)"
        }
        #if canImport(UIKit)
        return UIImage(cgImage: cgImage)
        #elseif canImport(AppKit)
        return NSImage(cgImage: cgImage, size: NSSize(width: cgImage.width, height: cgImage.height))
        #else
        return "VIPSImage(\(width)x\(height), \(bands) bands)"
        #endif
    }
}
